### ISSUE TITLE
Downstairs disconnect during LiveRepair can result in LiveRepair hanging forever

### ISSUE THREAD

**Issue by askfongjojo** [oxidecomputer/crucible#968]
I found a VM I had running for weeks throwing a bunch of EXT4-fs errors on the serial console:
<img width="1328" alt="ext4-fs-errors" src="https://github.com/oxidecomputer/crucible/assets/10874909/171b2f17-72eb-4634-b06b-114979928a2f">
The same problem happened to be reported by a customer today (the symptoms match but the root cause may differ).

The propolis log shows that it's waiting for a repair that is still going on:
```
BRM42220051 # zlogin oxz_propolis-server_38d396d5-eead-446b-a21f-b3c67d3c0479
[Connected to zone 'oxz_propolis-server_38d396d5-eead-446b-a21f-b3c67d3c0479' pts/13]
The illumos Project     helios-2.0.22187        September 2023
root@oxz_propolis-server_38d396d5-eead-446b-a21f-b3c67d3c0479:~# tail /var/svc/log/system-illumos-propolis-server\:default.log | looker
19:34:56.388Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Upstairs repair task running, trying again later
19:34:56.388Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair in progress, try again
19:35:56.392Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Checking if live repair is needed
19:35:56.392Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Upstairs repair task running, trying again later
19:35:56.392Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair in progress, try again
19:36:56.397Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Checking if live repair is needed
19:36:56.397Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Upstairs repair task running, trying again later
19:36:56.397Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair in progress, try again
```
But the downstairs serving the volume look healthy when I checked. One of them did report a repair that happened two days ago:
```
BRM44220005 # less /pool/ext/d1cb6b7d-2b92-4b7d-8a4d-551987f0277e/crypt/debug/oxz_crucible_f9c1deca-1898-429e-8c93-254c7aa7bae6/oxide-crucible-downstairs:downstairs-53c00db5-6deb-4b34-8862-68ec6a089b4a.log.1695663103 | looker
17:31:41.694Z WARN crucible: upstairs UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } disconnected, 0 jobs left
    task = main
17:31:41.694Z WARN crucible: upstairs UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } was previously active, clearing
    task = main
17:31:41.694Z INFO crucible: connection ([fd00:1122:3344:10b::23]:36777): all done
17:31:43.695Z INFO crucible: accepted connection from [fd00:1122:3344:10b::23]:46353
    task = main
17:31:43.695Z INFO crucible: Connection request from 94a61313-ccf2-469a-9614-81c667dae72f with version 4
    task = proc
17:31:43.695Z INFO crucible: upstairs UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } connected, version 4
    task = proc
17:31:43.696Z INFO crucible: UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } is now active (read-write)
17:31:43.709Z INFO crucible: Current flush_numbers [0..12]: [127808, 34823973, 34800951, 34800476, 32160233, 30788070, 30788070, 34795122, 23906888, 34776596, 33917646, 25085983]
17:31:43.738Z INFO crucible: Downstairs has completed Negotiation
    task = proc
```
I tried stopping and starting the VM to see if it is an instance of #815 but it isn't. The VM is stuck in `Stopping` state instead.
```
19:37:30.380Z INFO propolis-server: accepted connection
    local_addr = [fd00:1122:3344:10b::23]:12400
    remote_addr = [fd00:1122:3344:10b::1]:48086
19:37:30.400Z INFO propolis-server (vm_controller): Requested state Stop via API
19:37:30.403Z INFO propolis-server (external_request_queue): Queuing external request
    disposition = Enqueue
    request = Stop
19:37:30.404Z INFO propolis-server (vm_state_worker): State worker handling event
    event = External(Stop)
19:37:30.404Z INFO propolis-server (vm_state_worker): Stopping instance
19:37:30.404Z INFO propolis-server (vcpu_tasks): vCPU paused
    vcpu = 0
19:37:30.404Z INFO propolis-server (vcpu_tasks): vCPU paused
    vcpu = 1
19:37:30.405Z INFO propolis-server: request completed
    latency_us = 1285323156560
    local_addr = [fd00:1122:3344:10b::23]:12400
    method = GET
    remote_addr = [fd00:1122:3344:10b::1]:36957
    req_id = 4c1edf46-5f32-44b0-90b7-7c87d19816d1
    response_code = 200
    uri = /instance/state-monitor
19:37:30.406Z INFO propolis-server: request completed
    latency_us = 23482
    local_addr = [fd00:1122:3344:10b::23]:12400
    method = PUT
    remote_addr = [fd00:1122:3344:10b::1]:48086
    req_id = 16e15c93-4b05-4dfc-8ddf-3b41d22f8ae7
    response_code = 204
    uri = /instance/state
19:37:30.407Z INFO propolis-server (vm_controller): Sending pause request to lpc-bhyve-atpic
19:37:30.407Z INFO propolis-server (vm_controller): Sending pause request to lpc-bhyve-atpit
19:37:30.410Z INFO propolis-server (vm_controller): Sending pause request to lpc-bhyve-hpet
19:37:30.410Z INFO propolis-server (vm_controller): Sending pause request to lpc-bhyve-ioapic
19:37:30.410Z INFO propolis-server (vm_controller): Sending pause request to lpc-bhyve-rtc
19:37:30.410Z INFO propolis-server (vm_controller): Sending pause request to bhyve-vcpu-0
19:37:30.410Z INFO propolis-server (vm_controller): Sending pause request to bhyve-vcpu-1
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to chipset-i440fx
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to pci-piix4-hb
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to pci-piix3-lpc
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to pci-piix3-pm
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to lpc-bhyve-pmtimer
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to lpc-uart-com4
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to lpc-uart-com2
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to lpc-uart-com1
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to lpc-uart-com3
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to lpc-ps2ctrl
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to qemu-lpc-debug
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to pci-virtio-viona-0.9.0
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to pci-nvme-0.17.0
19:37:30.412Z INFO propolis-server (vm_controller): Sending pause request to block-crucible-94a61313-ccf2-469a-9614-81c667dae72f
19:37:56.401Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Checking if live repair is needed
19:37:56.401Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Upstairs repair task running, trying again later
19:37:56.401Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair in progress, try again
19:38:56.405Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Checking if live repair is needed
19:38:56.405Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Upstairs repair task running, trying again later
19:38:56.405Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair in progress, try again
```

**Comment by leftwo** [comment]
I think there is another issue that prevents propolis from stopping when crucible won't let go.
https://github.com/oxidecomputer/propolis/issues/371

It seems like either propolis can not wait on the pause, or we can make crucible pull the plug
on itself if a pause is received?  Or, maybe we want crucible to try and cleanly shutdown but give up an
pull the plug after X seconds?

**Comment by leftwo** [comment]
> But the downstairs serving the volume look healthy when I checked. One of them did report a repair that happened two days ago:

Those logs just tell me that the upstairs disconnected from the downstairs, then connected again.  I don't see any repair messages in that specific section.  If there is repair work done on a downstairs, there should be a bunch of messages that will indicate a repair has happened.

**Comment by askfongjojo** [comment]
Ah sorry, I pasted the wrong one. Here are the more relevant log lines which actually indicated that a repair failed:
```
02:12:39.957Z INFO crucible: Created copy dir "/data/regions/53c00db5-6deb-4b34-8862-68ec6a089b4a/00/000/001.copy"
02:12:40.014Z INFO crucible: eid:1 Found repair files: ["001", "001.db"]
02:12:45.302Z INFO crucible: Repair files downloaded, move directory "/data/regions/53c00db5-6deb-4b34-8862-68ec6a089b4a/00/000/001.copy" to "/data/regions/53c00db5-6deb-4b34-8862-68ec6a089b4a/00/000/001.replace"
02:12:45.304Z INFO crucible: Copy files from "/data/regions/53c00db5-6deb-4b34-8862-68ec6a089b4a/00/000/001.replace" in "/data/regions/53c00db5-6deb-4b34-8862-68ec6a089b4a/00/000"
02:12:45.461Z INFO crucible: Move directory  "/data/regions/53c00db5-6deb-4b34-8862-68ec6a089b4a/00/000/001.replace" to "/data/regions/53c00db5-6deb-4b34-8862-68ec6a089b4a/00/000/001.completed"
02:12:45.462Z WARN crucible: 174812934 job ELiveReopen for connection UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } waiting on 1 deps
    role = work
02:12:45.677Z WARN crucible: 174812954 job ELiveReopen for connection UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } waiting on 1 deps
    role = work
02:12:50.047Z WARN crucible: 174812964 job ELiveReopen for connection UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } waiting on 1 deps
    role = work
02:12:50.157Z WARN crucible: 174812986 job ELiveReopen for connection UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } waiting on 1 deps
    role = work
02:12:50.267Z WARN crucible: 174813002 job ELiveReopen for connection UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } waiting on 1 deps
    role = work
02:12:50.587Z WARN crucible: 174813021 job ELiveReopen for connection UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } waiting on 1 deps
    role = work
02:12:50.698Z INFO crucible: Created copy dir "/data/regions/53c00db5-6deb-4b34-8862-68ec6a089b4a/00/000/007.copy"
02:12:50.752Z ERRO crucible: connection ([fd00:1122:3344:10b::23]:46467) Exits with error: do_work_task task has ended: Ok(Err(Repair has failed, exiting task))
02:13:00.668Z INFO crucible: accepted connection from [fd00:1122:3344:10b::23]:43643
    task = main
02:13:00.668Z INFO crucible: Connection request from 94a61313-ccf2-469a-9614-81c667dae72f with version 4
    task = proc
02:13:00.668Z INFO crucible: upstairs UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } connected, version 4
    task = proc
02:13:00.668Z WARN crucible: Attempting RW takeover from UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } to UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 }
02:13:00.668Z WARN crucible: Signaling to UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } thread that UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } is being promoted (read-write)
02:13:00.668Z ERRO crucible: Error while signaling to UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } thread: SendError { .. }
02:13:00.668Z WARN crucible: Crucible Downstairs promoting UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } to active, discarding 3 jobs
02:13:00.669Z INFO crucible: UpstairsConnection { upstairs_id: 94a61313-ccf2-469a-9614-81c667dae72f, session_id: 86f0f8d6-4d4d-424d-8a5b-52189454cc90, gen: 2 } is now active (read-write)
02:13:00.685Z INFO crucible: Current flush_numbers [0..12]: [127808, 34823973, 34800951, 34800476, 32160233, 30788070, 30788070, 34795122, 23906888, 34776596, 33917646, 25085983]
02:13:00.706Z INFO crucible: Downstairs has completed Negotiation
```
Here is the complete log file for these log files:
[oxide-crucible-downstairs-downstairs-53c00db5-6deb-4b34-8862-68ec6a089b4a.log.1695446782.log](https://github.com/oxidecomputer/crucible/files/12743288/oxide-crucible-downstairs-downstairs-53c00db5-6deb-4b34-8862-68ec6a089b4a.log.1695446782.log)

I'll locate the one that has what triggered the repair shortly.

**Comment by askfongjojo** [comment]
The log attached in the comment above was already at the tip of the repair. Log files before it are empty and all subsequent logs have the same upstairs disconnected / connected again messages.

**Comment by leftwo** [comment]
> Ah sorry, I pasted the wrong one. Here are the more relevant log lines which actually indicated that a repair failed:

This does show a repair that was aborted, seemingly because the upstairs disconnected and then re-connected.
Which, still needs to be understood.

**Comment by leftwo** [comment]
Thanks to @askfongjojo and @bnaecker, I found the older propolis log files.
[system-illumos-propolis-server:default.log.1695437094.txt](https://github.com/oxidecomputer/crucible/files/12744054/system-illumos-propolis-server.default.log.1695437094.txt)

Of interest, we see the first time a downstairs (1) has a problem:
```
02:04:15.434Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] flow control start                                                                                                                                          
02:06:45.595Z WARN propolis-server (crucible-94a6
1313-ccf2-469a-9614-81c667dae72f): [up] downstairs 1 failed, too many outstanding jobs 57001                                                                                                       
02:06:45.600Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] client skip 57007 in process jobs because fault                                                                                                             
02:06:45.625Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] changed 57001 jobs to fault skipped                                                                                                                                                                                                                                                                                                   
```

So, we kicked the downstairs out because the queue depth was to large.

We see that the downstairs reconnects right away, then the repair starts:
```
02:11:35.744Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f Enter Ready for LiveRepair mode                                                                                        
02:11:35.746Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] Starts cmd_loop                                                                                                                                             
02:11:35.747Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] [fd00:1122:3344:105::8]:19000 task reports connection:true                                                                                                  02:11:35.747Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): 94a61313-ccf2-469a-9614-81c667dae72f Active LiveRepairReady Active                                                                                              
02:11:35.747Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Set check for repair                                                                                                                                            02:11:36.747Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Checking if live repair is needed                                                                                                                               
02:11:36.747Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f (86f0f8d6-4d4d-424d-8a5b-52189454cc90) Active LiveRepairReady Active ds_transition to LiveRepair                       
02:11:36.747Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] Transition from LiveRepairReady to LiveRepair                                                                                                               
02:11:36.747Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair started                                                                                                                                             
02:11:36.747Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair main task begins.                                                                                                                                                                                                                                                                                                                                                                         
02:11:36.747Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Start Live Repair of extents 0 to 1600                                                                                                                                                                                                        
02:11:36.747Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Start extent 0 repair 
```

Live repair chuggs along, and at extent 7 hits some trouble:
```
02:12:02.876Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Repair for extent 7 s:2 d:[1]                                                                                                                                                                                                                                                                                                                                     
02:12:02.876Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:7 Wait for result from repair command 174809908:174808909                                                                                                    
02:12:02.876Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174809908 final dependency list [174809906, 174809907]
02:12:03.248Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174809931 final dependency list [174809914] 
02:12:04.250Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174809934 final dependency list [174809931]                                                                                                                                                                                                                   
02:12:04.750Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174809938 final dependency list [174809934]                                                                                                                                                                                                      
02:12:14.771Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174809940 final dependency list [174809938]                                                                                                                                                                                
02:12:20.282Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174809942 final dependency list [174809940]
02:12:25.292Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174809945 final dependency list [174809942]                                                                                                                                                                                                   
02:12:26.795Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174809947 final dependency list [174809945]                                                                                                                                                                                         
02:12:28.680Z ERRO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] job id 174809908 saw error RepairStreamError("repair File { fd: 39 }: stream error: reqwest::Error { kind: Body, source: TimedOut }")                       
02:12:28.683Z ERRO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] DS Reports error Err(RepairStreamError("repair File { fd: 39 }: stream error: reqwest::Error { kind: Body, source: TimedOut }")) on job 174809908, DownstairsIO { ds_id: 174809908, guest_id: 174808909, work: ExtentLiveRepair { dependencies: [174809906, 174809907], extent: 7, source_downstairs: 2, source_repair_address: [fd00:1122:3344:106::5]:23001, repair_downstairs: [1] }, state: {0: Done, 1: InProgress, 2: Done}, ack_status: NotAcked, replay: false, data: None, read_response_hashes: [], impacted_blocks: InclusiveRange(ImpactedAddr { extent_id: 7, block: 0 }, ImpactedAddr { extent_id: 7, block: 131071 }) } EC                               
02:12:28.683Z ERRO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] Reports error RepairStreamError("repair File { fd: 39 }: stream error: reqwest::Error { kind: Body, source: TimedOut }") on job 174809908, DownstairsIO { d_id: 174809908, guest_id: 174808909, work: ExtentLiveRepair { dependencies: [174809906, 174809907], extent: 7, source_downstairs: 2, source_repair_address: [fd00:1122:3344:106::5]:23001, repair_downstairs: [1] }, state: {0: Done, 1: InProgress, 2: Done}, ack_status: NotAcked, replay: false, data: None, read_response_hashes: [], impacted_blocks: InclusiveRange(ImpactedAddr { extent_id: 7, block: 0 }, ImpactedAddr { extent_id: 7, block: 131071 }) }                                                                                                                                                                                                                            
02:12:28.683Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] client skip 40 in process jobs because fault                                 
02:12:28.683Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] changed 10 jobs to fault skipped 
02:12:28.683Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f (86f0f8d6-4d4d-424d-8a5b-52189454cc90) Active LiveRepair Active ds_transition to Faulted                               
02:12:28.683Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] Transition from LiveRepair to Faulted
```

Ahh, bad news, the repair failed on extent 7,
but, the show must go on.  LiveRepair now aborts the repair:

```
02:12:28.683Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] Transition from LiveRepair to Faulted                                                                                                                       
02:12:28.683Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] will exit pm_task, this downstairs Faulted                                                                                                                  
02:12:28.683Z ERRO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Extent 7 close id:174809908 Failed: IO Error: 1 out of 3 downstairs failed to complete this IO
02:12:28.683Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:7 Wait for result from NoOp command 174809909:174808910
02:12:28.684Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:7 Wait for result from reopen command 174809910:174808911
02:12:28.686Z ERRO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [fd00:1122:3344:105::8]:19000: proc: [1] client work task ended, Ok(Err([1] This downstairs now in Faulted)), so we end too
02:12:28.686Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f Gone missing, transition from Faulted to Faulted
02:12:28.686Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f connection to [fd00:1122:3344:105::8]:19000 closed
02:12:28.686Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] [fd00:1122:3344:105::8]:19000 task reports connection:false
02:12:28.686Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): 94a61313-ccf2-469a-9614-81c667dae72f Active Faulted Active
02:12:28.686Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] [fd00:1122:3344:105::8]:19000 task reports offline
02:12:28.787Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:7 Bailing with error
02:12:28.788Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Error After extent 7 repair
02:12:28.788Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Start extent 8 repair
02:12:28.788Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): extent 8 repair has failed
02:12:28.788Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Exit repair at extent 8
02:12:28.788Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): LiveRepair final flush submitted
02:12:28.790Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): LiveRepair final flush completed
02:12:28.790Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair returns Ok(())
```
(That "Live Repair returns Ok(())" might need some adjusted messaging.)

However, the crucible downstairs does not give up.  The upstairs soon reconnects to it:
```
02:12:35.196Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f looper connected
02:12:35.196Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] Proc runs for [fd00:1122:3344:105::8]:19000 in state Faulted
02:12:35.197Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] upstairs guest_io_ready=TRUE, promote! session 86f0f8d6-4d4d-424d-8a5b-52189454cc90
02:12:35.202Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] downstairs client at [fd00:1122:3344:105::8]:19000 has UUID 53c00db5-6deb-4b34-8862-68ec6a089b4a
```

Things start up and the Upstairs goes through the normal process of starting a repair:
```
02:12:35.242Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] [fd00:1122:3344:105::8]:19000 task reports connection:true
02:12:35.242Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): 94a61313-ccf2-469a-9614-81c667dae72f Active LiveRepairReady Active
02:12:35.242Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Set check for repair
02:12:36.243Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Checking if live repair is needed
02:12:36.243Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f (86f0f8d6-4d4d-424d-8a5b-52189454cc90) Active LiveRepairReady Active ds_transition to LiveRepair
02:12:36.243Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] Transition from LiveRepairReady to LiveRepair
02:12:36.243Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair started
02:12:36.243Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair main task begins.
02:12:36.243Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Start Live Repair of extents 0 to 1600 
02:12:36.243Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Start extent 0 repair
02:12:36.243Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:0 repair extent with ids 174812899,174812900,174812901,174812902 deps:[174812896]
02:12:36.243Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:0 close id:174812899 queued, notify DS
```

And, again, something bad happens while repairing extent 7 (and a key clue is here too, see if you can spot it)
```
02:12:50.588Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Start extent 7 repair
02:12:50.588Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:7 repair extent with ids 174813038,174813039,174813040,174813041 deps:[174813026]
02:12:50.588Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:7 close id:174813038 queued, notify DS
02:12:50.588Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:7 Wait for result from close command 174813038:174812039
02:12:50.588Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174813038 final dependency list [174813026]
02:12:50.588Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174813041 final dependency list [174813026, 174813038, 174813039, 174813040]
02:12:50.590Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174813044 final dependency list [174813041, 174813038, 174813026]
02:12:50.698Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Repair for extent 7 s:2 d:[1]
02:12:50.698Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:7 Wait for result from repair command 174813039:174812040
02:12:50.698Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174813039 final dependency list [174813026, 174813038]
02:12:50.844Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 174813051 final dependency list [174813044]
02:12:58.201Z ERRO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [fd00:1122:3344:105::8]:19000: proc: Connection reset by peer (os error 131)
02:12:58.201Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f Gone missing, transition from LiveRepair to Faulted
02:12:58.201Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f connection to [fd00:1122:3344:105::8]:19000 closed
02:12:58.201Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] [fd00:1122:3344:105::8]:19000 task reports connection:false
02:12:58.201Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): 94a61313-ccf2-469a-9614-81c667dae72f Active Faulted Active
02:12:58.201Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] [fd00:1122:3344:105::8]:19000 task reports offline
02:12:58.201Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] pm_task rx.recv() is None
```

This one looks like the downstairs disconnected from the upstairs, failing to finish work on extent 7.
Almost right away though, the upstairs reconnects to the downstairs:
```
02:13:00.668Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f looper connected
02:13:00.668Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] Proc runs for [fd00:1122:3344:105::8]:19000 in state Faulted
02:13:00.668Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] upstairs guest_io_ready=TRUE, promote! session 86f0f8d6-4d4d-424d-8a5b-52189454cc90
02:13:00.670Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] downstairs client at [fd00:1122:3344:105::8]:19000 has UUID 53c00db5-6deb-4b34-8862-68ec6a089b4a
```

And, after some other messages, we get to our stuck point:
```
02:13:00.707Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] 94a61313-ccf2-469a-9614-81c667dae72f Enter Ready for LiveRepair mode                                                                                        
02:13:00.707Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] Starts cmd_loop                                                                                                                                             
02:13:00.707Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): [1] [fd00:1122:3344:105::8]:19000 task reports connection:true                                                                                                  
02:13:00.707Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): 94a61313-ccf2-469a-9614-81c667dae72f Active LiveRepairReady Active                                                                                              
02:13:00.707Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Set check for repair                                                                                                                                            
02:13:01.707Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Checking if live repair is needed                                
02:13:01.707Z WARN propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Upstairs repair task running, trying again later                                                                                                                
02:13:01.707Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): Live Repair in progress, try again                                                                                                                              
```

Now, remember back earlier in the log, right before the downstairs went away this most recent time, spoiler
alert here:
```
02:12:50.698Z INFO propolis-server (crucible-94a61313-ccf2-469a-9614-81c667dae72f): RE:7 Wait for result from repair command 174813039:174812040
```
This takes us off into `wait_and_process_repair_ack()` which is what is waiting for the answer to a repair command.
My suspicion is the repair is stuck in that method, waiting for something to come over the `brw: block_req::BlockReqWaiter`, and nothing is.

This makes sense, as I don't see any log messages that come after this method, and that a new repair is being
blocked from starting

**Comment by leftwo** [comment]
For tracking down where things are,
This is for the disk attached to the instance in question
```
root@oxz_switch:/tmp# ./omdb db disks info 94a61313-ccf2-469a-9614-81c667dae72f
HOST_SERIAL DISK_NAME         INSTANCE_NAME     PROPOLIS_ZONE                                            VOLUME_ID                            
BRM42220051 sysbench-mysql-13 sysbench-mysql-13 oxz_propolis-server_38d396d5-eead-446b-a21f-b3c67d3c0479 0d0ada85-82ba-46c4-b94e-d0c3c1b28c05 
HOST_SERIAL REGION                               ZONE                                              PHYSICAL_DISK                        
BRM42220031 e49f6d97-620f-4009-acd8-051352c59738 oxz_crucible_2b34cd1d-ea7d-41a1-82b9-75550fdf6eb0 a83bdc49-daf2-4248-b238-fda4239305de 
BRM42220006 2834134e-0ddb-42d8-8fa9-818438d12733 oxz_crucible_46d1afcc-cc3f-4b17-aafc-054dd4862d15 3f6ca632-29ad-41c8-86ae-13c77caf5b92 
BRM44220005 53c00db5-6deb-4b34-8862-68ec6a089b4a oxz_crucible_f9c1deca-1898-429e-8c93-254c7aa7bae6 6a19ddca-447e-4579-90b3-6826fa3f7950 
```

**Comment by leftwo** [comment]
With the new rewrite, the conditions that created this issue no longer exist!

The fix is in https://github.com/oxidecomputer/crucible/pull/1058
